### ISSUE TITLE
Proper Use Of Plugin In Android (Example provided by me) should be in documentation

### ISSUE THREAD

**Issue by b-cancel** [Baseflow/flutter-permission-handler#147]
## 🏗 Enhancement Proposal

NOTE: only Android, I have no idea the different edge cases that need to be covered on IOS

Many people have often, in one form or another, mentioned 2 things

1. In the first run, when calling "requestPermissions" it returns PermissionStatus.denied when it should instead logically return PermissionStatus.unknown.

You have explained: that there is nothing you can do to improve this given how android functions.

2. In the first run, when calling "shouldShowRequestPermissionRationale" what is returned is not accurate 

Others have explained: that until you first at least call "requestPermissions" first the result may or may not be accurate

Both of the above are just what is mentioned but there are more edge cases to cover and no proper way has been suggested to cover them anywhere on your documentation

### Pitch

Since I am getting an app ready for production I have taken the time to code and thoroughly test a solution and I propose that you add this to the documentation so others can have an easier time at using this without having to learn the ins and outs of how your plugin functions

```
//request contact permission from the user
import 'package:flutter/material.dart';
import 'package:permission_handler/permission_handler.dart';

//request contact permission from the user
requestPermission({
   //the SINGLE permission you are requesting
  @required PermissionGroup thePermission,
  //whether or not you are forcing the user to grant you permission
  @required bool forcePermission,
  //what happens when the permission is given
  @required Function onPermissionGiven,
  //what happens when the permission isn't given (denied, or hard denied)
  @required Function onPermissionNotGiven,
  //Function that brings up another 1. Page or 2. Pop Up now that the standard permission request won't come up
  @required Function manualRequest,
}) async{
   //NOTE: we know for a fact that the permission status here the FIRST TIME around IS NOT authorized
  //PermissionStatus status = (await Permission.getPermissionsStatus([PermissionName.Contacts]))[0].permissionStatus;
  PermissionStatus status = await PermissionHandler().checkPermissionStatus(thePermission);
  if(isStatusAuthorized(status)) onPermissionGiven();
  else{ //NOTE: we know the negative status is VALID BUT...
    //NOTE: at the moment "shouldShowRequestPermissionRationale" is not going to be accurate in all cases
    //it becomes accurate in all cases once you atleast request the permission first
    bool firstDontShowRationale = await PermissionHandler().shouldShowRequestPermissionRationale(thePermission);
    bool secondDontShowRationale;

    //request permission
    //NOTE: this MAY just be skipped over in cases where the user has previously told the app to never request the permission again
    Map<PermissionGroup, PermissionStatus> perm = await PermissionHandler().requestPermissions([thePermission]).then((var p)async{
      secondDontShowRationale = await PermissionHandler().shouldShowRequestPermissionRationale(thePermission);
      return p;
    });

    //if the pop up came up the user may have allowed the permission
    PermissionStatus status = perm[thePermission];
    if(isStatusAuthorized(status)) onPermissionGiven();
    else{
      //NOTE: in all cases 
      //1. approving the permission allways works as expected
      //2. not deciding what to do when being soft requested by closing the app
      //    - keep the permission in its current state

      //CASES (firstDontShowRationale | secondDontShowRationale)
      //  -First Request did bringing up the pop up-
      //1. [false and true] on the absolute first time requesting the permission
      //2. [true and true] after the first soft denial and it continues to be the case until the last soft denial
      //3. [true and false] after the first permanent denial (only happens once)
      //  -First Request did not bringing up the pop up-
      //4. [false and false] after the from then on

      //from the observation above we can determine whether or not the request pop up came up and what to do about it
      bool popUpCameUp = secondDontShowRationale || (firstDontShowRationale && secondDontShowRationale == false);
      bool canSoftRequestAgain = secondDontShowRationale;

      //if the pop up came up react to the decision of the user
      if(popUpCameUp){
        //if we are not forcing permission AND we either denied
        //NOTE: denied happens with either SOFT denied or HARD denied
        if(forcePermission == false && status == PermissionStatus.denied){ 
          //pops back to the page that requested this contact change
          onPermissionNotGiven();
        }
        else{
          //continue asking
          if(canSoftRequestAgain){
            requestPermission(
              thePermission: thePermission,
              forcePermission: forcePermission, 
              onPermissionGiven: onPermissionGiven, 
              onPermissionNotGiven: onPermissionNotGiven,
              manualRequest: manualRequest,
            );
          }
          else manualRequest();
        }
      } 
      else manualRequest(); 
    }
  }
}
```

### Platforms affected (mark all that apply)
- [ ] :iphone: iOS
- [X] :robot: Android

**Comment by b-cancel** [comment]
@martijn00 I went hard on the comments so any other developers using this can know exactly what is being done and why

I hope this helps many others

**Comment by b-cancel** [comment]
I refined things even further

```
//request contact permission from the user
requestPermission({
   //the SINGLE permission you are requesting
  @required PermissionGroup thePermission,
  //whether or not you are forcing the user to grant you permission
  @required bool forcePermission,
  //what happens when the permission is given
  @required Function onPermissionGiven,
  //what happens when the permission isn't given (denied, or hard denied)
  @required Function onPermissionNotGiven,
  //Function that brings up another 1. Page or 2. Pop Up now that the standard permission request won't come up
  @required Function manualRequest,
}) async{
  //NOTE: we know for a fact that the permission status here the FIRST TIME around IS NOT authorized
  //PermissionStatus status = (await Permission.getPermissionsStatus([PermissionName.Contacts]))[0].permissionStatus;
  PermissionStatus status = await PermissionHandler().checkPermissionStatus(thePermission);
  if(isStatusAuthorized(status)) onPermissionGiven();
  else{ //NOTE: we know the negative status is VALID BUT...
    //NOTE: at the moment "shouldShowRequestPermissionRationale" is not going to be accurate in all cases
    //it becomes accurate in all cases once you atleast request the permission first
    bool firstDontShowRationale = await PermissionHandler().shouldShowRequestPermissionRationale(thePermission);
    bool secondDontShowRationale;

    //request permission
    //NOTE: this MAY just be skipped over in cases where the user has previously told the app to never request the permission again
    Map<PermissionGroup, PermissionStatus> perm = await PermissionHandler().requestPermissions([thePermission]).then((var p)async{
      secondDontShowRationale = await PermissionHandler().shouldShowRequestPermissionRationale(thePermission);
      return p;
    });

    //NOTE: in all cases 
    //1. approving the permission allways works as expected
    //2. not deciding what to do when being soft requested by closing the app
    //    - keep the permission in its current state

    //CASES (firstDontShowRationale | secondDontShowRationale)
    //  -First Request did bringing up the pop up-
    //1. [false and true] on the absolute first time requesting the permission
    //2. [true and true] after the first soft denial and it continues to be the case until the last soft denial
    //3. [true and false] after the first permanent denial (only happens once)
    //  -First Request did not bringing up the pop up-
    //4. [false and false] after the from then on

    //from the observation above we can determine whether or not the request pop up came up and what to do about it
    bool popUpCameUp = secondDontShowRationale || (firstDontShowRationale && secondDontShowRationale == false);
    bool canSoftRequestAgain = secondDontShowRationale;

    //if the pop up came up react to the decision of the user
    if(popUpCameUp){
      PermissionStatus status = perm[thePermission];
      if(isStatusAuthorized(status)) onPermissionGiven();
      else{
        //if we are not forcing permission AND we either denied
        //NOTE: denied happens with either SOFT denied or HARD denied
        if(forcePermission == false && status == PermissionStatus.denied){ 
          //pops back to the page that requested this contact change
          onPermissionNotGiven();
        }
        else{
          //continue asking
          if(canSoftRequestAgain){
            requestPermission(
              thePermission: thePermission,
              forcePermission: forcePermission, 
              onPermissionGiven: onPermissionGiven, 
              onPermissionNotGiven: onPermissionNotGiven,
              manualRequest: manualRequest,
            );
          }
          else manualRequest();
        }
      }
    } //the pop up didn't come up so its not possible for the status to have changed
    else manualRequest();    
  }
}
```

**Comment by cremor** [comment]
I agree that `shouldShowRequestPermissionRationale` needs more documentation, but I don't think that such a complex sample helps very much. A better starting point would be something like the sample shown in the [official Android documentation](https://developer.android.com/training/permissions/requesting#make-the-request).

Also, why do you mix `await` and `.then()` in your code?

**Comment by b-cancel** [comment]
@cremor this example is indeed complex but when you remove all the comments it becomes quite breif (ill do so below)

the reason why I await and .then() is because I need to wait for the pop up to either come up 
or not come up

It will come up if the user has never actually said "dont pop up again"
it will not if the user has said that

Ideally shouldShowRequestPermissionRationale does this
but because of the limitation it does not
so the work around is the only option

I've tested it extensively
feel free to test it for yourself

here is the simplified code

```
requestPermission({
  @required PermissionGroup thePermission,
  @required bool forcePermission,
  @required Function onPermissionGiven,
  @required Function onPermissionNotGiven,
  @required Function manualRequest,
}) async{
  PermissionStatus status = await PermissionHandler().checkPermissionStatus(thePermission);
  if(isStatusAuthorized(status)) onPermissionGiven();
  else{ 
    bool firstDontShowRationale = await PermissionHandler().shouldShowRequestPermissionRationale(thePermission);
    bool secondDontShowRationale;

    //request permission
    //NOTE: this MAY just be skipped over in cases where the user has previously told the app to never request the permission again
    Map<PermissionGroup, PermissionStatus> perm = await PermissionHandler().requestPermissions([thePermission]).then((var p)async{
      secondDontShowRationale = await PermissionHandler().shouldShowRequestPermissionRationale(thePermission);
      return p;
    });

    //if the pop up came up the user may have allowed the permission
    PermissionStatus status = perm[thePermission];
    if(isStatusAuthorized(status)) onPermissionGiven();
    else{
      bool popUpCameUp = secondDontShowRationale || (firstDontShowRationale && secondDontShowRationale == false);
      bool canSoftRequestAgain = secondDontShowRationale;

      //if the pop up came up react to the decision of the user
      if(popUpCameUp){
        //if we are not forcing permission AND they denied us permission
        //NOTE: denied happens with either SOFT denied or HARD denied
        if(forcePermission == false && status == PermissionStatus.denied){ 
          onPermissionNotGiven();
        }
        else{
          //continue asking
          if(canSoftRequestAgain){
            requestPermission(
              thePermission: thePermission,
              forcePermission: forcePermission, 
              onPermissionGiven: onPermissionGiven, 
              onPermissionNotGiven: onPermissionNotGiven,
              manualRequest: manualRequest,
            );
          }
          else manualRequest();
        }
      } 
      else manualRequest(); 
    }
  }
}
```

**Comment by cremor** [comment]
> the reason why I await and .then() is because I need to wait for the pop up to either come up
> or not come up
> 
> It will come up if the user has never actually said "dont pop up again"
> it will not if the user has said that

You mean that the `Future` returned from `shouldShowRequestPermissionRationale` never finishes if the user has chosen "don't show again"? That's not intuitive at all and I would consider it a bug.

**Comment by b-cancel** [comment]
@cremor nope,
the Future completes without the pop up coming up IF the user has chosen "don't show again"
the Future completes with the pop up coming up IF the user has chosen "deny" or "allow"

in either case, you need to request first to get the actual valid return value of "shouldShowRequestPermissionRationale"

Note how it's possible for it to change without user action (Cases explained below)
The First boolean is firstDontShowRationale
The Second boolean is secondDontShowRationale

either of these being true means -> show the rationale
either of these being false means -> don't show the rationale

```CASE1: [false and true] on the absolute first time requesting the permission```
1. first-time function says -> yes show rationale! 
when really the user has not yet said "don't show again"
so if you request the pop up it will come up
so ideally the function would instead say -> don't show rationale

* This is because of the android limitation that has been mentioned before

& I imagine this is the case because perhaps android wants you to explain to your users why the permission is being requested before you request it

The problem with this is that now we need to do some weird code (like what I have above) to determine whether we really NEED to show rationale

NOTE: how the function is called "SHOULD show rationale" and not "NEED to show rationale"

so I guess technically it works fine... but it's less helpful for the developer
showing the rationale at its core is a design decision but I have no idea why it's being forced on us

after the permission is requested once regularly then the program assumes that the rationale was displayed

its all rather wonky really

but the code I wrote lets people not worry about all that and just get the thing to work

2. second-time function says -> whatever the user selected on the pop-up

Let me know if you need me to explain the cases below

```CASE 2: [true and true] after the first soft denial and it continues to be the case until the last soft denial```
```CASE 3: [true and false] after the first permanent denial (only happens once)```
```CASE 4: [false and false] after all other permanent denials```

**Comment by b-cancel** [comment]
@cremor many updates made above, perhaps your end merits a refresh
but hopefully, I'm getting my point across

**Comment by wreyford** [comment]
where is isStatusAuthorized(status) defined in your example code

**Comment by mvanbeusekom** [comment]
This issue will be solved in our next release (4.2.0), where on Android the methods `checkPermissionStatus` and `requestPermissions` will return the `neverAskAgain` permission status when the user has denied the permission and selected the "Don't ask again" checkbox.

More info in PR #189.
